### PR TITLE
 console: Select the first embeddable graphics device as graphical console

### PR DESCRIPTION
**Pull request by flynnjiang**
Currently, when there are multiple graphics devices, the first graphics
device is always selected as the graphical console. For the following
configuration:

    <graphics type="egl-headless">
      <gl rendernode="/dev/dri/renderD128"/>
    </graphics>
    <graphics type="spice" autoport="yes">
      <listen type="address"/>
      <image compression="off"/>
      <gl enable="no"/>
    </graphics>

virt-manager will report an error:

    Cannot display graphical console type 'egl-headless'

The patch will iterate through all graphics devices to try to select the
first embeddable graphics device as graphical console.

### CODE REVIEW

File: virtManager/config.py [hunk 1]
```diff
@@ -212,6 +212,17 @@ def embeddable_graphics(self):
...
+    def get_perfered_graphics_index(self, gdevs):
+        try:
+            for idx, dev in enumerate(gdevs):
+                if dev.type in self.embeddable_graphics():
+                    return idx
+            if len(gdevs) > 0:
+                return 0
+            return None
+        except:
+            return None
```
Comment (crobinso) [inline virtManager/config.py:224]: This whole function, and the `embeddable_graphics` function, can be moved into console.py actually. `embeddable_graphics` is only called there, it's a remnant from earlier days. Moving 'embeddable_graphics` could be a prep patch

File: virtManager/config.py [hunk 2]
```diff
@@ -212,6 +212,17 @@ def embeddable_graphics(self):
...
 
+    def get_perfered_graphics_index(self, gdevs):
+        try:
+            for idx, dev in enumerate(gdevs):
+                if dev.type in self.embeddable_graphics():
+                    return idx
+            if len(gdevs) > 0:
+                return 0
+            return None
+        except:
```
Comment (crobinso) [inline virtManager/config.py:223]: This except block should be removed. if we screw up the basic code inside, we want to know about it

File: virtManager/config.py [hunk 3]
```diff
@@ -212,6 +212,17 @@ def embeddable_graphics(self):
         ret = ["vnc", "spice"]
         return ret
 
+    def get_perfered_graphics_index(self, gdevs):
```
Comment (crobinso) [inline virtManager/config.py:215]: spelling is `preferred`

File: virtManager/details/viewers.py [hunk 4]
```diff
@@ -134,7 +134,7 @@ def _get_fd_for_open(self):
         if not self._vm.conn.support.domain_open_graphics():
             return None
 
-        return self._vm.open_graphics_fd()
+        return self._vm.open_graphics_fd(self._ginfo.gidx)
```
Comment (crobinso) [inline virtManager/details/viewers.py:137]: All this gidx handling could go into the previous patch. Also kill the idx=0 default param in the first patch, force the one user to make it explicit.

File: virtManager/details/console.py [hunk 5]
```diff
@@ -911,7 +917,7 @@ def _console_list_menu_toggled(self, src, dev):
 
     def _populate_console_list_menu(self, ignore=None):
         self._consolemenu.rebuild_menu(
-                self.vm, self._console_list_menu,
+                self.vm, self.config, self._console_list_menu,
```
Comment (crobinso) [inline virtManager/details/console.py:920]: All this passing `self.config` around shouldn't be required when you move the functions out of config.py. Maybe they would need to be made global or something

**Review by crobinso — COMMENTED** [review]
Thanks, this looks pretty good! I've got some small comments

File: virtManager/config.py (hunk 1 above)
Comment (flynnjiang) [reply virtManager/config.py]: Thanks very much for your suggestions! I have moved `embeddable_graphics` to console.py and removed `get_prefered_grahpics_index`  in v2  after some optimization.

File: virtManager/config.py (hunk 2 above)
Comment (flynnjiang) [reply virtManager/config.py]: This function was removed.

File: virtManager/details/viewers.py (hunk 4 above)
Comment (flynnjiang) [reply virtManager/details/viewers.py:137]: I've fixed it. Thanks very much!

File: virtManager/config.py (hunk 3 above)
Comment (flynnjiang) [reply virtManager/config.py]: This function was removed. Thanks very much!

File: virtManager/details/console.py (hunk 5 above)
Comment (flynnjiang) [reply virtManager/details/console.py]: I have fixed it.

**Comment by crobinso** [comment]
Sorry for the very late response. Code still mostly applies so I gave it a spin.

There's a problem though. There's already some kind of plumbing for listing graphical multiple graphical connections, View->Consoles. However that has the assumption baked in that we only ever try to connect to the first listed graphical device. We need to resolve that somehow, otherwise the UI is out of sync. If I get some time soon i'll try to straighten it out.

**Comment by crobinso** [comment]
> Sorry for the very late response. Code still mostly applies so I gave it a spin.
> 
> There's a problem though. There's already some kind of plumbing for listing graphical multiple graphical connections, View->Consoles. However that has the assumption baked in that we only ever try to connect to the first listed graphical device. We need to resolve that somehow, otherwise the UI is out of sync. If I get some time soon i'll try to straighten it out.

Scratch that, I see you tried to handle it! My testing was bad.

**Comment by crobinso** [comment]
Pushed now, thanks! I'm adding some other cleanups and refactorings on top, if you get some time please confirm your usecase is still working correctly with latest git

**Comment by flynnjiang** [comment]
Your help is greatly appreciated! I just test it with the latest commit (a6b62a20b62ac86e0c40d18f188b312513bafdd9), and it works.

**Pull request merged**